Sources/Web3Core/KeystoreManager/BIP44.swift

```diff
@@ -0,0 +1,126 @@
+//
+//  BIP44.swift
+//  Created by Alberto Penas Amor on 15/12/22.
+//
+
+import Foundation
+
+public protocol BIP44 {
+    /**
+    Derive an ``HDNode`` based on the provided path. The function will throw ``BIP44Error.warning`` if it was invoked with throwOnWarning equal to
+    `true` and the root key doesn't have a previous child with at least one transaction. If it is invoked with throwOnError equal to `false` the child node will be
+    derived directly using the derive function of ``HDNode``. This function needs to query the blockchain history when throwOnWarning is `true`, so it can throw
+    network errors.
+    - Parameter path: valid BIP44 path.
+    - Parameter throwOnWarning: `true` to use
+    [Account Discovery](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#account-discovery) standard,
+    otherwise it will dervive the key using the derive function of ``HDNode``.
+    - Throws: ``BIP44Error.warning`` if the child key shouldn't be used according to
+    [Account Discovery](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#account-discovery) standard.
+    - Returns: an ``HDNode`` child key for the provided `path` if it can be created, otherwise `nil`
+    */
+    func derive(path: String, throwOnWarning: Bool, transactionChecker: TransactionChecker) async throws -> HDNode?
+}
+
+public enum BIP44Error: LocalizedError, Equatable {
+    /// The selected path doesn't fulfill BIP44 standard, you can derive the root key anyway
+    case warning
+
+    public var errorDescription: String? {
+        switch self {
+        case .warning:
+            return "Couldn't derive key as it doesn't have a previous account with at least one transaction"
+        }
+    }
+}
+
+public protocol TransactionChecker {
+    /**
+    It verifies if the provided address has at least one transaction
+    - Parameter address: to be queried
+    - Throws: any error related to query the blockchain provider
+    - Returns: `true` if the address has at least one transaction, `false` otherwise
+    */
+    func hasTransactions(address: String) async throws -> Bool
```

Comment:
Let's change the type of address to `EthereumAddress`.
This way we reduce one potential error - an invalid ethereum address.

The fact that we implement `TransactionChecker` and before we call `hasTransactions` function we check that the address is valid we do not provide safety for others if they want to implement a version of `TransactionChecker` themselves.

Any thoughts are welcome!